lib/proxy.js

```diff
@@ -590,65 +620,79 @@ ReverseProxy.buildRoute = function (route) {
   return routeObject;
 };
 
-ReverseProxy.prototype._getTarget = function (src, req) {
+ReverseProxy.prototype._getTarget = function(src, req, res) {
   var url = req.url;
 
-  return this.resolve(src, url, req).bind(this).then(function (route) {
-    if (!route) {
-      this.log && this.log.warn({ src: src, url: url }, 'no valid route found for given source');
-      return;
-    }
+  return this.resolve(src, url, req)
+    .bind(this)
+    .then(function(route) {
+      if (!route) {
+        this.log && this.log.warn({ src: src, url: url }, 'no valid route found for given source');
+        return;
+      }
+
+      var pathname = route.path;
+      if (pathname.length > 1) {
+        //
+        // remove prefix from src
+        //
+        req._url = url; // save original url
+        req.url = url.substr(pathname.length) || '';
+      }
 
-    var pathname = route.path;
-    if (pathname.length > 1) {
       //
-      // remove prefix from src
+      // Perform Round-Robin on the available targets
+      // TODO: if target errors with EHOSTUNREACH we should skip this
+      // target and try with another.
       //
-      req._url = url; // save original url
-      req.url = url.substr(pathname.length) || '/';
-    }
+      var urls = route.urls;
+      var j = route.rr;
+      route.rr = (j + 1) % urls.length; // get and update Round-robin index.
+      var target = route.urls[j];
 
-    //
-    // Perform Round-Robin on the available targets
-    // TODO: if target errors with EHOSTUNREACH we should skip this
-    // target and try with another.
-    //
-    var urls = route.urls;
-    var j = route.rr;
-    route.rr = (j + 1) % urls.length; // get and update Round-robin index.
-    var target = route.urls[j];
+      //
+      // Fix request url if targetname specified.
+      //
+      if (target.pathname) {
+        if (req.url) {
+          req.url = path.posix.join(target.pathname, req.url);
+        } else {
+          req.url = target.pathname;
+        }
+      }
 
-    //
-    // Fix request url if targetname specified.
-    //
-    if (target.pathname) {
-      req.url = path.join(target.pathname, req.url);
-    }
+      //
+      // Host headers are passed through from the source by default
+      // Often we want to use the host header of the target instead
+      //
+      if (target.useTargetHostHeader === true) {
+        req.host = target.host;
+      }
 
-    //
-    // Host headers are passed through from the source by default
-    // Often we want to use the host header of the target instead
-    //
-    if (target.useTargetHostHeader === true) {
-      req.host = target.host;
-    }
+      this.log &&
+        this.log.info('Proxying %s to %s', src + url, path.posix.join(target.host, req.url));
```

Comment:
this log becomes incorrect since it shows the target.host however maybe we are goint to proxy to x